File: 2024 June/Check If two Line segments Intersect/main.py
# Link: https://www.geeksforgeeks.org/problems/check-if-two-line-segments-intersect0017/1


class Solution:
    def doIntersect(self, p1, q1, p2, q2):
        m1 = (q1[1] - p1[1]) / (q1[0] - p1[0])
        m2 = (q2[1] - p2[1]) / (q2[0] - p2[0])
        if m1 == m2:
            return "false"

        chk211 = (p2[1] - p1[1]) - m1 * (p2[0] - p1[0]) > 0
        chk212 = (q2[1] - p1[1]) - m1 * (q2[0] - p1[0]) > 0
        chk121 = (p1[1] - p2[1]) - m2 * (p1[0] - p2[0]) > 0
        chk122 = (q1[1] - p2[1]) - m2 * (q1[0] - p2[0]) > 0

        if (
            (chk211 and chk212)
            or (not chk211 and not chk212)
            or (chk121 and chk122)
            or (not chk121 and not chk122)
        ):
            return "false"
        return "true"


if __name__ == "__main__":
    obj = Solution()
    print(obj.doIntersect([1, 1], [10, 1], [1, 2], [10, 2]))
    print(obj.doIntersect([10, 0], [0, 10], [0, 0], [10, 10]))

File: 2024 June/Check If two Line segments Intersect/main.cpp
// Link: https://www.geeksforgeeks.org/problems/check-if-two-line-segments-intersect0017/1

#include <iostream>
using namespace std;

class Solution
{
public:
    string doIntersect(int p1[], int q1[], int p2[], int q2[])
    {
        double m1 = (double)(q1[1] - p1[1]) / (double)(q1[0] - p1[0]);
        double m2 = (double)(q2[1] - p2[1]) / (double)(q2[0] - p2[0]);
        if (m1 == m2)
        {
            return "false";
        }

        bool chk211 = p2[1] - p1[1] - m1 * (p2[0] - p1[0]) > 0 ? true : false;
        bool chk212 = q2[1] - p1[1] - m1 * (q2[0] - p1[0]) > 0 ? true : false;
        bool chk121 = p1[1] - p2[1] - m2 * (p1[0] - p2[0]) > 0 ? true : false;
        bool chk122 = q1[1] - p2[1] - m2 * (q1[0] - p2[0]) > 0 ? true : false;

        if ((chk211 == true && chk212 == true) || (chk211 == false && chk212 == false) ||
            (chk121 == true && chk122 == true) || (chk121 == false && chk122 == false))
        {
            return "false";
        }
        return "true";
    }
};

int main()
{
    Solution obj;
    int p11[] = {1, 1};
    int q11[] = {10, 1};
    int p21[] = {1, 2};
    int q21[] = {10, 2};
    cout << obj.doIntersect(p11, q11, p21, q21) << endl;

    int p12[] = {10, 0};
    int q12[] = {0, 10};
    int p22[] = {0, 0};
    int q22[] = {10, 10};
    cout << obj.doIntersect(p12, q12, p22, q22) << endl;

    return 0;
}